### PR TITLE
Respect `--dart-sdk` parameter when running analyze.dart tests.

### PR DESCRIPTION
**Pull request by stereotype441**
The `--dart-sdk` parameter is passed in when invoking this script from
the Dart team's `flutter-analyze` bot, which is intended to provide an
early warning if a commit to the Dart SDK would cause problems with
the Dart SDK roll into Flutter (see
https://github.com/dart-lang/sdk/blob/main/tools/bots/flutter/analyze_flutter_flutter.sh).
This change ensures that when `analyze.dart` is run by that bot, and
it invokes `dart` or `pub`, it uses the version of Dart that's under
test, rather than the version of Dart that's been rolled into Flutter
sources.

This change should allow us to simplify the Dart repository's
`analyze_flutter_flutter.sh` script, which we believe will
eliminate the flakiness we're currently seeing with the
`flutter-analyze` bot.

https://github.com/dart-lang/sdk/issues/47881 tracks the
flakiness of the `flutter-analyze` bot.

*List which issues are fixed by this PR. You must list at least one issue.*

- https://github.com/dart-lang/sdk/issues/47881 (not fully fixed by this PR; additional work will need to be done on the Dart side after this PR lands)

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

This reminds me that we should probably go back to using the args package in this script, now that it is null-safe. No need to do that for this PR.

File: dev/bots/analyze.dart [hunk 1]
```diff
@@ -25,8 +25,9 @@ final String flutterRoot = path.dirname(path.dirname(path.dirname(path.fromUri(P
 final String flutter = path.join(flutterRoot, 'bin', Platform.isWindows ? 'flutter.bat' : 'flutter');
 final String flutterPackages = path.join(flutterRoot, 'packages');
 final String flutterExamples = path.join(flutterRoot, 'examples');
-final String dart = path.join(flutterRoot, 'bin', 'cache', 'dart-sdk', 'bin', Platform.isWindows ? 'dart.exe' : 'dart');
-final String pub = path.join(flutterRoot, 'bin', 'cache', 'dart-sdk', 'bin', Platform.isWindows ? 'pub.bat' : 'pub');
+late final String dartSdk;
+final String dart = path.join(dartSdk, 'bin', Platform.isWindows ? 'dart.exe' : 'dart');
```
Comment (Hixie) [inline dev/bots/analyze.dart:29]: isn't this guaranteed to fail?

Comment (Hixie) [reply dev/bots/analyze.dart:29]: (if it's not guaranteed to fail it's very confusing code)

File: dev/bots/analyze.dart [hunk 2]
```diff
@@ -45,6 +50,20 @@ Future<void> main(List<String> arguments) async {
   print('$clock ${bold}Analysis successful.$reset');
 }
 
+/// Scans [arguments] for an argument of the form `--dart-sdk` or
+/// `--dart-sdk=...` and returns the configured SDK, if any.
+String? _getDartSdkFromArguments(List<String> arguments) {
+  for (int i = 0; i < arguments.length; i++) {
```
Comment (Hixie) [inline dev/bots/analyze.dart:56]: trivial nit: prefer `+= 1` (see style guide)

**Comment by Hixie** [comment]
This actually is testable, we have several tests for this script, see the `test` subdirectory. It's not strictly required to test this script per our policy, but it might be _worth_ testing, given that if it breaks the failure mode is to silently use the wrong SDK, which might take a very long time to detect and meanwhile might cause all manner of issues. I'll let you decide whether to test it or not.

test-exempt: is a test

File: dev/bots/analyze.dart (hunk 1 above)
Comment (stereotype441) [reply dev/bots/analyze.dart:29]: All static variables with initializers are lazily evaluated, so this works fine provided that `main` sets `dartSdk` before any attempt is made to evaluate `dart` or `pub`.

What would help to clear up your confusion?  Would you like me to mark `dart` and `pub` as `late final`?  Restructure the code in some way?

File: dev/bots/analyze.dart [hunk 3]
```diff
@@ -45,6 +50,20 @@ Future<void> main(List<String> arguments) async {
...
+    if (arguments[i] == '--dart-sdk' && i + 1 < arguments.length) {
+      return arguments[i + 1];
```
[the remark this answers is not in the document]
Comment (stereotype441) [reply dev/bots/analyze.dart]: Done

File: dev/bots/analyze.dart (hunk 1 above)
Comment (Hixie) [reply dev/bots/analyze.dart]: Yeah that's much less confusing, thanks.

**Pull request merged**